File: step_definitions/Verify Real-Time Reflection of Changes on Dashboard.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import LoginPage from "../pages/loginPage";
import DashboardPage from "../pages/dashboardPage";

Given("the user is on the login page", () => {
  LoginPage.visit();
});

When("the user logs in with valid credentials", () => {
  LoginPage.enterUsername('validUsername');
  LoginPage.enterPassword('validPassword');
  LoginPage.submit();
});

Then("the user is successfully logged in and directed to the dashboard", () => {
  DashboardPage.verifyDashboard();
});

Given("the user is logged in", () => {
  LoginPage.visit();
  LoginPage.enterUsername('validUsername');
  LoginPage.enterPassword('validPassword');
  LoginPage.submit();
  DashboardPage.verifyDashboard();
});

When("the user navigates to the dashboard section", () => {
  DashboardPage.navigateToSection();
});

Then("the dashboard is displayed with all components visible", () => {
  DashboardPage.verifyAllComponentsVisible();
});

Given("the user is on the dashboard", () => {
  DashboardPage.verifyDashboard();
});

When("the user selects a component to edit", () => {
  DashboardPage.selectComponentToEdit();
});

Then("the component is selected and options to edit are displayed", () => {
  DashboardPage.verifyEditOptionsDisplayed();
});

Given("the user has selected a component to edit", () => {
  DashboardPage.selectComponentToEdit();
});

When("the user makes changes to the component", () => {
  DashboardPage.makeChangesToComponent();
});

Then("the changes are made successfully with no errors", () => {
  DashboardPage.verifyChangesMade();
});

Given("changes have been made to a dashboard component", () => {
  DashboardPage.makeChangesToComponent();
  DashboardPage.verifyChangesMade();
});

When("the user saves the changes", () => {
  DashboardPage.saveChanges();
});

Then("the changes are saved successfully without any errors", () => {
  DashboardPage.verifyChangesSaved();
});

When("the user observes the dashboard", () => {
  DashboardPage.observeDashboard();
});

Then("the dashboard updates in real-time to reflect the changes made", () => {
  DashboardPage.verifyRealTimeUpdates();
});

When("the user refreshes the dashboard page", () => {
  DashboardPage.refreshPage();
});

Then("the dashboard retains the changes after refresh", () => {
  DashboardPage.verifyChangesRetainedAfterRefresh();
});

When("the user logs out and logs back into the application", () => {
  DashboardPage.logout();
  LoginPage.visit();
  LoginPage.enterUsername('validUsername');
  LoginPage.enterPassword('validPassword');
  LoginPage.submit();
});

Then("the dashboard still reflects the changes after re-login", () => {
  DashboardPage.verifyDashboard();
  DashboardPage.verifyChangesRetainedAfterReLogin();
});

When("the user checks the dashboard on a different device/browser", () => {
  DashboardPage.checkOnDifferentDevice();
});

Then("the changes are visible across different devices/browsers in real-time", () => {
  DashboardPage.verifyChangesAcrossDevices();
});

Given("the user is logged in on two different devices", () => {
  DashboardPage.loginOnMultipleDevices();
});

When("the user makes simultaneous changes on both devices", () => {
  DashboardPage.makeSimultaneousChanges();
});

Then("real-time sync is maintained and changes are reflected accurately", () => {
  DashboardPage.verifyRealTimeSync();
});

When("the user checks for error messages or logs", () => {
  DashboardPage.checkForErrors();
});

Then("no error messages or logs indicate issues with real-time updates", () => {
  DashboardPage.verifyNoErrors();
});

Given("the dashboard is receiving multiple real-time updates", () => {
  DashboardPage.receiveMultipleUpdates();
});

When("the user observes the dashboard performance", () => {
  DashboardPage.observePerformance();
});

Then("the dashboard performance remains optimal", () => {
  DashboardPage.verifyPerformance();
});

Given("the user wants to test different types of changes", () => {
  DashboardPage.prepareForDifferentChanges();
});

When("the user makes various changes", () => {
  DashboardPage.makeVariousChanges();
});

Then("all types of changes are reflected in real-time without issues", () => {
  DashboardPage.verifyAllChanges();
});

Given("the user has a slow internet connection", () => {
  DashboardPage.simulateSlowConnection();
});

When("the user makes changes to the dashboard", () => {
  DashboardPage.makeChangesToComponent();
});

Then("real-time updates are consistent even with a slow internet connection", () => {
  DashboardPage.verifyRealTimeWithSlowConnection();
});

When("the user reverts the changes", () => {
  DashboardPage.revertChanges();
});

Then("reverted changes are reflected in real-time accurately", () => {
  DashboardPage.verifyRevertedChanges();
});

File: step_definitions/Verify Real-Time Reflection of Changes on Dashboard.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import com.page_objects.DashboardPage;

public class DashboardSteps extends DashboardPage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user is on the login page")
    public void theUserIsOnTheLoginPage() {
        launchUrl(getAppUrl());
        maximizeWindow();
    }

    @When("the user logs in with valid credentials")
    public void theUserLogsInWithValidCredentials() {
        String userName = testHarness.getData("LoginData", "UserName");
        String password = testHarness.getData("LoginData", "Password");
        enterUserName(userName);
        enterPassword(password);
        clickLoginButton();
    }

    @Then("the user is successfully logged in and directed to the dashboard")
    public void theUserIsSuccessfullyLoggedInAndDirectedToTheDashboard() {
        verifyDashboardPageRedirection();
    }

    @Given("the user is logged in")
    public void theUserIsLoggedIn() {
        theUserIsOnTheLoginPage();
        theUserLogsInWithValidCredentials();
    }

    @When("the user navigates to the dashboard section")
    public void theUserNavigatesToTheDashboardSection() {
        navigateToDashboard();
    }

    @Then("the dashboard is displayed with all components visible")
    public void theDashboardIsDisplayedWithAllComponentsVisible() {
        verifyDashboardComponentsVisibility();
    }

    @Given("the user is on the dashboard")
    public void theUserIsOnTheDashboard() {
        theUserIsLoggedIn();
        navigateToDashboard();
    }

    @When("the user selects a component to edit")
    public void theUserSelectsAComponentToEdit() {
        selectDashboardComponent();
    }

    @Then("the component is selected and options to edit are displayed")
    public void theComponentIsSelectedAndOptionsToEditAreDisplayed() {
        verifyEditOptionsDisplayed();
    }

    @Given("the user has selected a component to edit")
    public void theUserHasSelectedAComponentToEdit() {
        theUserIsOnTheDashboard();
        selectDashboardComponent();
    }

    @When("the user makes changes to the component \(e.g., update data, add widget\)")
    public void theUserMakesChangesToTheComponent() {
        makeChangesToComponent();
    }

    @Then("the changes are made successfully with no errors")
    public void theChangesAreMadeSuccessfullyWithNoErrors() {
        verifyChangesMadeSuccessfully();
    }

    @Given("changes have been made to a dashboard component")
    public void changesHaveBeenMadeToADashboardComponent() {
        theUserHasSelectedAComponentToEdit();
        makeChangesToComponent();
    }

    @When("the user saves the changes")
    public void theUserSavesTheChanges() {
        saveComponentChanges();
    }

    @Then("the changes are saved successfully without any errors")
    public void theChangesAreSavedSuccessfullyWithoutAnyErrors() {
        verifyChangesSavedSuccessfully();
    }

    @When("the user observes the dashboard")
    public void theUserObservesTheDashboard() {
        observeDashboard();
    }

    @Then("the dashboard updates in real-time to reflect the changes made")
    public void theDashboardUpdatesInRealTimeToReflectTheChangesMade() {
        verifyRealTimeUpdates();
    }

    @When("the user refreshes the dashboard page")
    public void theUserRefreshesTheDashboardPage() {
        refreshDashboardPage();
    }

    @Then("the dashboard retains the changes after refresh")
    public void theDashboardRetainsTheChangesAfterRefresh() {
        verifyChangesRetainedAfterRefresh();
    }

    @When("the user logs out and logs back into the application")
    public void theUserLogsOutAndLogsBackIntoTheApplication() {
        logoutAndLoginAgain();
    }

    @Then("the dashboard still reflects the changes after re-login")
    public void theDashboardStillReflectsTheChangesAfterReLogin() {
        verifyChangesAfterReLogin();
    }

    @When("the user checks the dashboard on a different device\/browser")
    public void theUserChecksTheDashboardOnADifferentDeviceBrowser() {
        checkDashboardOnDifferentDevice();
    }

    @Then("the changes are visible across different devices\/browsers in real-time")
    public void theChangesAreVisibleAcrossDifferentDevicesBrowsersInRealTime() {
        verifyChangesAcrossDevices();
    }

    @Given("the user is logged in on two different devices")
    public void theUserIsLoggedInOnTwoDifferentDevices() {
        loginOnMultipleDevices();
    }

    @When("the user makes simultaneous changes on both devices")
    public void theUserMakesSimultaneousChangesOnBothDevices() {
        makeSimultaneousChanges();
    }

    @Then("real-time sync is maintained and changes are reflected accurately")
    public void realTimeSyncIsMaintainedAndChangesAreReflectedAccurately() {
        verifyRealTimeSync();
    }

    @When("the user checks for error messages or logs")
    public void theUserChecksForErrorMessagesOrLogs() {
        checkForErrorMessages();
    }

    @Then("no error messages or logs indicate issues with real-time updates")
    public void noErrorMessagesOrLogsIndicateIssuesWithRealTimeUpdates() {
        verifyNoErrorMessages();
    }

    @Given("the dashboard is receiving multiple real-time updates")
    public void theDashboardIsReceivingMultipleRealTimeUpdates() {
        simulateMultipleRealTimeUpdates();
    }

    @When("the user observes the dashboard performance")
    public void theUserObservesTheDashboardPerformance() {
        observeDashboardPerformance();
    }

    @Then("the dashboard performance remains optimal")
    public void theDashboardPerformanceRemainsOptimal() {
        verifyOptimalPerformance();
    }

    @Given("the user wants to test different types of changes")
    public void theUserWantsToTestDifferentTypesOfChanges() {
        prepareForDifferentChanges();
    }

    @When("the user makes various changes \(e.g., data update, visual change\)")
    public void theUserMakesVariousChanges() {
        makeVariousChanges();
    }

    @Then("all types of changes are reflected in real-time without issues")
    public void allTypesOfChangesAreReflectedInRealTimeWithoutIssues() {
        verifyAllChangesReflected();
    }

    @Given("the user has a slow internet connection")
    public void theUserHasASlowInternetConnection() {
        simulateSlowInternetConnection();
    }

    @When("the user makes changes to the dashboard")
    public void theUserMakesChangesToTheDashboard() {
        makeChangesWithSlowConnection();
    }

    @Then("real-time updates are consistent even with a slow internet connection")
    public void realTimeUpdatesAreConsistentEvenWithASlowInternetConnection() {
        verifyRealTimeUpdatesWithSlowConnection();
    }

    @When("the user reverts the changes")
    public void theUserRevertsTheChanges() {
        revertDashboardChanges();
    }

    @Then("reverted changes are reflected in real-time accurately")
    public void revertedChangesAreReflectedInRealTimeAccurately() {
        verifyRevertedChanges();
    }
}